From 954567abfcd410041adc473956490a49ff42c5bb Mon Sep 17 00:00:00 2001
From: Peter Martienssen <peter.martienssen@liquify-consulting.de>
Date: Fri, 13 Aug 2021 18:58:14 +0200
Subject: [PATCH 9/9] Added a control 'trigger_mode' to activate one of 7
 different trigger modes of the VC MIPI Cameras. Additionaly added two
 controls 'image_height' and 'vmax' for faster testing while driver is in
 development.

---
 .../platform/tegra/camera/tegracam_ctrls.c    | 86 +++++++++++++++++++
 kernel/nvidia/include/media/camera_common.h   |  3 +
 .../nvidia/include/media/tegra-v4l2-camera.h  |  3 +
 3 files changed, 92 insertions(+)

diff --git a/kernel/nvidia/drivers/media/platform/tegra/camera/tegracam_ctrls.c b/kernel/nvidia/drivers/media/platform/tegra/camera/tegracam_ctrls.c
index 4bca039d4..99468af34 100644
--- a/kernel/nvidia/drivers/media/platform/tegra/camera/tegracam_ctrls.c
+++ b/kernel/nvidia/drivers/media/platform/tegra/camera/tegracam_ctrls.c
@@ -174,6 +174,36 @@ static struct v4l2_ctrl_config ctrl_cfg_list[] = {
 		.max = STEREO_EEPROM_SIZE,
 		.step = 2,
 	},
+	{
+		.ops = &tegracam_ctrl_ops,
+		.id = TEGRA_CAMERA_CID_TRIGGER_MODE,
+		.name = "Trigger Mode",
+		.type = V4L2_CTRL_TYPE_INTEGER64,
+		.flags = V4L2_CTRL_FLAG_SLIDER,
+		.min = 0,
+		.max = 7,
+		.step = 1,
+	},
+	{
+		.ops = &tegracam_ctrl_ops,
+		.id = TEGRA_CAMERA_CID_IMAGE_HEIGHT,
+		.name = "Image Height",
+		.type = V4L2_CTRL_TYPE_INTEGER64,
+		.flags = V4L2_CTRL_FLAG_SLIDER,
+		.min = 0,
+		.max = 10000,
+		.step = 1,
+	},
+	{
+		.ops = &tegracam_ctrl_ops,
+		.id = TEGRA_CAMERA_CID_VMAX,
+		.name = "Vmax",
+		.type = V4L2_CTRL_TYPE_INTEGER64,
+		.flags = V4L2_CTRL_FLAG_SLIDER,
+		.min = 0,
+		.max = 10000,
+		.step = 1,
+	},
 };
 
 static int tegracam_get_ctrl_index(u32 cid)
@@ -307,6 +337,15 @@ static int tegracam_set_ctrls(struct tegracam_ctrl_handler *handler,
 
 	/* For controls that require sensor to be on */
 	switch (ctrl->id) {
+	case TEGRA_CAMERA_CID_TRIGGER_MODE:
+		err = ops->set_trigger_mode(tc_dev, *ctrl->p_new.p_s64);
+		break;
+	case TEGRA_CAMERA_CID_IMAGE_HEIGHT:
+		err = ops->set_image_height(tc_dev, *ctrl->p_new.p_s64);
+		break;
+	case TEGRA_CAMERA_CID_VMAX:
+		err = ops->set_vmax(tc_dev, *ctrl->p_new.p_s64);
+		break;
 	case TEGRA_CAMERA_CID_GAIN:
 		err = ops->set_gain(tc_dev, *ctrl->p_new.p_s64);
 		break;
@@ -696,6 +735,28 @@ static int tegracam_check_ctrl_ops(
 			else
 				string_ops++;
 			break;
+		case TEGRA_CAMERA_CID_TRIGGER_MODE:
+			if (ops->set_trigger_mode == NULL)
+				dev_err(dev,
+					"Missing TEGRA_CAMERA_CID_TRIGGER_MODE implementation\n");
+			if (ops->set_trigger_mode != NULL)
+				sensor_ops++;
+			break;
+		case TEGRA_CAMERA_CID_IMAGE_HEIGHT:
+			if (ops->set_image_height == NULL)
+				dev_err(dev,
+					"Missing TEGRA_CAMERA_CID_IMAGE_HEIGHT implementation\n");
+			if (ops->set_image_height != NULL)
+				sensor_ops++;
+			break;
+		case TEGRA_CAMERA_CID_VMAX:
+			if (ops->set_vmax == NULL)
+				dev_err(dev,
+					"Missing TEGRA_CAMERA_CID_VMAX implementation\n");
+			if (ops->set_vmax != NULL)
+				sensor_ops++;
+			break;
+
 		case TEGRA_CAMERA_CID_FUSE_ID:
 			if (tegracam_get_string_ctrl_size(
 					TEGRA_CAMERA_CID_FUSE_ID, ops) == 0)
@@ -855,6 +916,31 @@ static int tegracam_check_ctrl_cids(struct tegracam_ctrl_handler *handler)
 			errors_found++;
 		}
 	}
+	if (ops->set_trigger_mode != NULL) {
+		if (!find_matching_cid(ops->ctrl_cid_list,
+			ops->numctrls,
+			TEGRA_CAMERA_CID_TRIGGER_MODE)) {
+			dev_err(dev, "Missing TEGRA_CAMERA_CID_TRIGGER_MODE registration\n");
+			errors_found++;
+		}
+	}
+	if (ops->set_image_height != NULL) {
+		if (!find_matching_cid(ops->ctrl_cid_list,
+			ops->numctrls,
+			TEGRA_CAMERA_CID_IMAGE_HEIGHT)) {
+			dev_err(dev, "Missing TEGRA_CAMERA_CID_IMAGE_HEIGHT registration\n");
+			errors_found++;
+		}
+	}
+	if (ops->set_vmax != NULL) {
+		if (!find_matching_cid(ops->ctrl_cid_list,
+			ops->numctrls,
+			TEGRA_CAMERA_CID_VMAX)) {
+			dev_err(dev, "Missing TEGRA_CAMERA_CID_VMAX registration\n");
+			errors_found++;
+		}
+	}
+
 
 	/* Find missing string control IDs */
 	if (ops->fill_string_ctrl != NULL) {
diff --git a/kernel/nvidia/include/media/camera_common.h b/kernel/nvidia/include/media/camera_common.h
index d2d7906c2..857d85952 100644
--- a/kernel/nvidia/include/media/camera_common.h
+++ b/kernel/nvidia/include/media/camera_common.h
@@ -191,6 +191,9 @@ struct tegracam_ctrl_ops {
 	u32 compound_ctrl_size[TEGRA_CAM_MAX_COMPOUND_CONTROLS];
 	const u32 *ctrl_cid_list;
 	bool is_blob_supported;
+	int (*set_trigger_mode)(struct tegracam_device *tc_dev, s64 val);
+	int (*set_image_height)(struct tegracam_device *tc_dev, s64 val);
+	int (*set_vmax)(struct tegracam_device *tc_dev, s64 val);
 	int (*set_gain)(struct tegracam_device *tc_dev, s64 val);
 	int (*set_exposure)(struct tegracam_device *tc_dev, s64 val);
 	int (*set_exposure_short)(struct tegracam_device *tc_dev, s64 val);
diff --git a/kernel/nvidia/include/media/tegra-v4l2-camera.h b/kernel/nvidia/include/media/tegra-v4l2-camera.h
index 619218d70..a4cc008ca 100644
--- a/kernel/nvidia/include/media/tegra-v4l2-camera.h
+++ b/kernel/nvidia/include/media/tegra-v4l2-camera.h
@@ -38,6 +38,9 @@
 #define TEGRA_CAMERA_CID_FRAME_RATE		(TEGRA_CAMERA_CID_BASE+11)
 #define TEGRA_CAMERA_CID_EXPOSURE_SHORT		(TEGRA_CAMERA_CID_BASE+12)
 #define TEGRA_CAMERA_CID_STEREO_EEPROM		(TEGRA_CAMERA_CID_BASE+13)
+#define TEGRA_CAMERA_CID_TRIGGER_MODE		(TEGRA_CAMERA_CID_BASE+14)
+#define TEGRA_CAMERA_CID_IMAGE_HEIGHT		(TEGRA_CAMERA_CID_BASE+15)
+#define TEGRA_CAMERA_CID_VMAX			(TEGRA_CAMERA_CID_BASE+16)
 
 #define TEGRA_CAMERA_CID_SENSOR_CONFIG		(TEGRA_CAMERA_CID_BASE+50)
 #define TEGRA_CAMERA_CID_SENSOR_MODE_BLOB	(TEGRA_CAMERA_CID_BASE+51)
-- 
2.25.1

